misc: Enable Spark query runner as reference in aggregation fuzzer test

rui-mo opened the pull request:
- Enables Spark query runner as reference DB in Spark aggregation fuzzer test. 
- Removes duplicate Spark aggregation fuzzer test from experimental run.
- Implements `executeAndReturnVector` in Spark query runner and removes 
the below two APIs from the reference query runner as they are not in use since
https://github.com/facebookincubator/velox/commit/8519fad1170975ab32ee75e186a88f91d3098495.

```
virtual std::multiset<std::vector<velox::variant>> execute(
      const std::string& /*sql*/,
      const std::vector<velox::RowVectorPtr>& /*input*/,
      const velox::RowTypePtr& /*resultType*/) {
    VELOX_UNSUPPORTED();
  }

virtual std::vector<RowVectorPtr> executeVector(
      const std::string& /*sql*/,
      const std::vector<RowVectorPtr>& /*input*/,
      const RowTypePtr& /*resultType*/) {
    VELOX_UNSUPPORTED();
  }
```

- Removes the Spark source file download. Instead, photo files are introduced 
in Velox and need to be updated for every Spark version. Currently, these files
are imported from Spark 3.5.1 version.

Fixes https://github.com/facebookincubator/velox/issues/9270.

majetideepak commented on velox/exec/tests/CMakeLists.txt:
```diff
@@ -186,7 +194,96 @@ target_link_libraries(
+  utf8::utf8_range
+  utf8::utf8_validity
+  utf8::utf8_range_lib
+  absl::absl_kernel_timeout_internal
```
Is protobuf or re2 pulling in absl?

rui-mo replied:
In my test environment, the protobuf installed by grpc pulls in absl.

rui-mo commented:
Hi @mbasmanova, this is a draft PR to enable Spark query runner in aggregate fuzzer test. I'd like to work on below items in separate PRs. Could you spare some time to check if they make sense? Thank you.

1. https://github.com/facebookincubator/velox/pull/9759
2. Refactor the `toSql`  process by introducing QueryTranslator class to allow reusability for Spark when queries are the same.
3. Fix the intermediate type of Sum aggregate discovered when running fuzzer test.
4. Introduce a configuration to specify if the vectors generated by VectorFuzzer are for Arrow parquet. We need it because exportFlattenedVector method in Bridge.cpp does not support nested encoding and non-scalar type, which would breaks parquet write.
5. Introduce SparkQueryRunner and enable it in agg fuzzer test.

mbasmanova commented:
@rui-mo Rui, thank you for working on this. At a high-level next steps make sense to me, but there are not enough details for me to understand them fully.

> Fix the intermediate type of Sum aggregate discovered when running fuzzer test.

Would you create a GitHub issue to describe this problem?

> Introduce a configuration to specify if the vectors generated by VectorFuzzer are for Arrow parquet. We need it because exportFlattenedVector method in Bridge.cpp does not support nested encoding and non-scalar type, which would breaks parquet write.

Does this mean we won't be able to create tables with map/array/struct columns? If so, this is a pretty severe limitation.

rui-mo commented:
@mbasmanova Thanks for your reply.
For the sum aggregate issue, I notice https://github.com/facebookincubator/velox/pull/9818 is addressing the same issue as discovered by this fuzzer test.
For the second one, I created https://github.com/facebookincubator/velox/issues/9821 and provided some details. Thanks.

rui-mo commented:
@mbasmanova @majetideepak Would you like to take a review? Thanks.

majetideepak commented on .github/workflows/experimental.yml:
```diff
@@ -180,37 +180,55 @@ jobs:
-    runs-on: ubuntu-latest
-    needs: compile
+  spark-java-aggregation-fuzzer-run:
+    runs-on: 16-core-ubuntu
```
@assignUser FYI

majetideepak reviewed (commented):
@rui-mo CPP code changes look good.
@assignUser can you please take another look at the workflow file changes? Thanks.

assignUser commented on .github/workflows/scheduled.yml:
```diff
@@ -454,9 +454,9 @@ jobs:
+    container: ghcr.io/facebookincubator/velox-dev:spark-server
     needs: compile
-    timeout-minutes: 60
+    timeout-minutes: 120
```
Why is this change necessary? The usual duration is still 15/30mins. Did you actually see a timeout?

rui-mo replied:
Got it. This change is not necessary and just reverted it.

assignUser commented on .github/workflows/experimental.yml:
If we are adding this to `scheduled.yml` do we still need it here?

rui-mo commented on .github/workflows/experimental.yml:
```diff
@@ -85,18 +85,6 @@ jobs:
-        uses: actions/upload-artifact@v3
-        with:
-          name: aggregation
-          path: velox/_build/debug/velox/functions/prestosql/fuzzer/velox_aggregation_fuzzer_test
```
To remove the duplicate one with L86.

rui-mo commented:
@assignUser Fixed above comments. Would you like to take another look? Thanks.

assignUser reviewed (approved):
Thanks!

mbasmanova commented:
@rui-mo Would you rebase to allow merging?

rui-mo commented:
Hi @kevinwilfong, could you help import and merge this PR? Thanks!

kgpai commented:
Not sure why this is blocked. 
@rui-mo Can you fix the conflict , I will work to land this as soon as thats fixed.

kgpai commented:
@rui-mo I see a compilation failure when building fuzzer jobs, can you investigate/fix ?

rui-mo commented:
@kgpai Thanks a lot for your help. I will investigate. I also notice there is an API change in the fuzzer test, and I'm incorporating this PR with the updated code.

rui-mo commented:
Hi @kgpai, I added this change in the Common.h to fix the compilation error, but it seems the experimental test is for the main branch so it cannot be verified (I'm not sure why this error only occurs in this PR). Do you have any suggestion?
> using dwio::common::StreamIdentifier::operator==;

```
/home/runner/work/velox/velox/velox/./velox/dwio/common/StreamIdentifier.h:40:16: error: ‘virtual bool facebook::velox::dwio::common::StreamIdentifier::operator==(const facebook::velox::dwio::common::StreamIdentifier&) const’ was hidden [-Werror=overloaded-virtual=]
   40 |   virtual bool operator==(const StreamIdentifier& other) const {
      |                ^~~~~~~~
/home/runner/work/velox/velox/velox/./velox/dwio/dwrf/common/Common.h:203:8: note:   by ‘bool facebook::velox::dwrf::DwrfStreamIdentifier::operator==(const facebook::velox::dwrf::DwrfStreamIdentifier&) const’
  203 |   bool operator==(const DwrfStreamIdentifier& other) const {
      |        ^~~~~~~~
```

rui-mo commented:
This PR is updated, and please check the PR description for further information. Thanks!

kgpai commented:
@rui-mo The experimental build is triggered because the experimental.yml has changed. Can you raise an issue regarding that. We can leave fixing that outside the scope of this change. Thanks !

rui-mo commented:
@kgpai I opened an issue https://github.com/facebookincubator/velox/issues/12342 regarding the compilation failure. Thanks!

kgpai commented:
Hi @rui-mo , Wanted to give you an update, this requires some changes internally before I can land this. Please give me some time.

rui-mo commented:
@kgpai Many thanks for your help! Regarding the status of this PR, the compilation failure is being fixed by https://github.com/facebookincubator/velox/pull/12356, and one Spark aggregate result issue was found in CI and is being fixed by https://github.com/facebookincubator/velox/pull/12335.

kgpai commented:
@rui-mo Theres unfortunately a problem with this PR. With this change spark aggregation fuzzer now depends on the spark proto libraries. This works in cmake obviously because of resolve dependencies which allows us to pull files like 'spark/connect/base.grpc.pb.h' out - however internally at meta we do not have spark (nor is it advisable to pull all of spark) as a third party dependency. I realize that the dependency on spark protobufs is not from this PR though - the previous PR that added them and was landed did not have a target against SparkQueryRunner and thus we were able to land it. 

I think the way forward is to add 'spark/connect/base.grpc.pb.h' etc as external dependencies in velox/external and then use them - this way it should work internally/externally ; Unfortunately though it means we will have to keep updating this for every version of spark etc. 

Let me know what you think @rui-mo  ? Feel free to reach out to me on slack to discuss this further.

rui-mo commented:
@kgpai Thank you very much for the help! I understand the problem here. I suppose we can either add 'spark/connect/base.grpc.pb.h' etc as external dependencies like you mentioned, or add the proto files to the 'sparksql/fuzzer' folder and compile the `*.pb.h`  and `*.pb.cc` during each compilation. With the latter way less files would be introduced to Velox, and in the future the compilation of them might only happen when 'VELOX_BUILD_FUZZERS' is enabled, seeing https://github.com/facebookincubator/velox/pull/10802.  How do you think?

kgpai commented:
@rui-mo Yes adding the proto files as a external dependency to be compiled into the pb files is the better approach. Would we need to ensure that the protobuf version is same for Spark / Velox going forward ?

rui-mo commented:
@kgpai I suppose the protobuf library versions used by the client and server do not need to match exactly, but they need to be compatible. Spark uses proto 3 which should be compatible with Velox-used version.

rui-mo commented:
@kgpai I updated this PR to introduce the photo files of Spark and removed the Spark downloading process. Could you help check if the internal issue in Meta can be resolved? Thanks for your help.

assignUser reviewed (changes requested):
Please note that the ASL 2.0 requires that we add the content of Spark's NOTICE file to our own. You can check how we have previously attributed ASF projects there.

rui-mo commented:
@assignUser Sorry, I did not get it. I find [orc_proto.proto](https://github.com/facebookincubator/velox/blob/main/velox/dwio/dwrf/proto/orc_proto.proto) also uses this license, and the Spark proto files are copied from [spark/connect](https://github.com/apache/spark/tree/v3.5.1/connector/connect/common/src/main/protobuf/spark/connect). Would you mind giving me further details on what I need to do? Thanks.

rui-mo commented:
@assignUser Thanks for the suggestion. I updated this PR to add Spark's NOTICE file to Velox NOTICE.txt. Could you help take another review?

https://github.com/facebookincubator/velox/pull/9559/files#diff-b810ebe4ffda4293c79d292059b45b59de7e649e86314ffd5a1ee4eb4b1e08c8R113

assignUser reviewed (approved):
Thanks for updating the notice file!

rui-mo commented:
Thanks you all for helping review! @kgpai Could you help import this PR if no further comment? The compilation issue of experimental test could be fixed by https://github.com/facebookincubator/velox/pull/12854.

rui-mo commented:
The CI failure is irrelevant to the change of this PR, and opened https://github.com/facebookincubator/velox/issues/13144 to track.

kgpai commented:
@rui-mo Please give me a few days to work on landing this..

kgpai commented on velox/functions/sparksql/fuzzer/spark/connect/catalog.proto:
```diff
@@ -0,0 +1,243 @@
+
+package spark.connect;
+
+import "spark/connect/common.proto";
```
@rui-mo All these paths need to be relative to velox , ie "velox/functions/sparksql/fuzzer/spark/connect/common.proto" etc.. 
I cannot get it to build internally without that..
Can you make that change ?
